Additional tasks for module 8

TatyanaV24 opened the pull request:
Дополнительные задания.
Пока выгрузила только первую задачу и то, она решена частично. Надо довести до ума кнопку стоп, на сколько я понимаю она должна срабатывать, как кнопка пауза.
Остальные задачи добавлю постепенно.

TatyanaV24 committed:
.

TatyanaV24 committed:
Задача 1 из доп.заданий

1. Для галереи картинок (пункт 2 про 10 картинок) добавить кнопку старт и стоп. По нажатию на старт начинают проигрываться анимировано все картинки.  По нажатию на стоп останавливается. Останавливать необходимо на той картинке, которая показывалась в момент нажатия на стоп .
Картинки анимированно проигрываются и останавливаются по нажатию на стоп, при этом сразу отображается первая картирнка, а не та на которой была нажата кнопка стоп.

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/ViewController.swift:
```diff
@@ -0,0 +1,24 @@
+    }
+    @IBAction func stopButton(_ sender: UIButton) {
+        playImageView.stopAnimating()
+    }
```
👍

Встроенными средствами UIImageView нельзя удовлетворить все требования с анимацией картинок одновременно.
Попробуете воспользоваться Timer?
https://developer.apple.com/documentation/foundation/timer

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/ViewController.swift:
```diff
@@ -0,0 +1,24 @@
+
+class ViewController: UIViewController {
+    
+    var catsImage = [UIImage(named: "1")!, UIImage(named: "2")!,UIImage(named: "3")!, UIImage(named: "4")!, UIImage(named: "5")!, UIImage(named: "6")!, UIImage(named: "7")!, UIImage(named: "8")!, UIImage(named: "9")!, UIImage(named: "10")!]
```
еще один способ создания:
`let catsImage = ["1", "2", "3", "4"].compactMap { UIImage(named: $0) }`

ios-review replied:
такая запись немного короче

ios-review reviewed (changes requested):
Решать задачу постепенно?
Мне нравится такая идея)

Добавил комментарии.

TatyanaV24 committed:
Доп задачи по блоку 8

Дмитрий, старалась решить как можно большее количество дополнительных заданий по блоку 8.
Я вставила условия задач в самом вверху контролеров с кодом.
Конечно без вопросов не обойдется.
Первый вопрос по задаче 3, надо было вставить картинки в кнопки для UIStepper. Основываясь на документации, все выполнено верно, но реализация выходит очень кривой, сами изображения не отображаются, просто стоят синие квадратики. Буду признательна, если поясните, возможно я допустила ошибку.
Еще вопрос по задаче 5, ее я не решила в полном объеме (даже контроллер не стала добавлять) сбилась с логикой действий. Я правильно понимаю, в UIDataPicker - устанавливаем нужную дату, в UITextField - часовой пояс в виде числа, UIlabel - должен отображать дату с учетом нового часового пояса?

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/ProgressViewController.swift:
```diff
@@ -0,0 +1,47 @@
+        } else if progressView.progress == 1.0 {
+            progressView.progress = 0.0
+        }
+    }
```
create переводится как создать.
В функции  же выполняется нечто другое.
Какое название будет более подходящим, более полно описывающим происходящее?

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/ProgressViewController.swift:
```diff
@@ -0,0 +1,47 @@
+        } else {
+            self.myTime.invalidate()
+        }
+    }
```
По нажатию на кнопки play и stop нужно производить противоположные действия.

В этом случае нагляднее разделить функции на два @IBAction.

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/SlaiderViewController.swift:
```diff
@@ -0,0 +1,34 @@
+    @IBAction func changeImageSlider(_ sender: UISlider) {
+        index = Int(round(sender.value))
+    }
+}
```
Все задания с галерей картинок можно реализовать на одном экране.
По сути будут просто разные способы изменения текущей картинки на экране.

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/SlaiderViewController.swift:
```diff
@@ -0,0 +1,34 @@
+
+import UIKit
+
+class SlaiderViewController: UIViewController {
```
Ошибка layout
<img width="1438" alt="Screenshot 2021-01-24 at 16 28 18" src="https://user-images.githubusercontent.com/62285448/105626367-b8312800-5e61-11eb-9801-18d8416137c4.png">

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/StepperViewController.swift:
```diff
@@ -0,0 +1,26 @@
+        customStepper.setDecrementImage(UIImage(named: "s-2"), for: .normal)
+        customStepper.translatesAutoresizingMaskIntoConstraints = false
+        
+    }
```
Ничего не происходит по нажатию на клавиши степпера.
<img width="501" alt="Screenshot 2021-01-24 at 16 33 29" src="https://user-images.githubusercontent.com/62285448/105626410-f595b580-5e61-11eb-902f-67cd45435893.png">

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/ProgressViewController.swift:
```diff
@@ -0,0 +1,47 @@
+
+import UIKit
+
+class ProgressViewController: UIViewController {
```
Ошибка layout
<img width="1434" alt="Screenshot 2021-01-24 at 16 34 39" src="https://user-images.githubusercontent.com/62285448/105626435-1f4edc80-5e62-11eb-9e78-37eaf4295dd5.png">

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/SwitchViewController.swift:
```diff
@@ -0,0 +1,23 @@
+
+import UIKit
+
+class SwitchViewController: UIViewController {
```
На этом экране тоже есть ошибка layout для SE (2nd).

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/ActivityIndicatorViewController.swift:
```diff
@@ -0,0 +1,31 @@
+
+import UIKit
+
+class ActivityIndicatorViewController: UIViewController {
```
И тут ошибка layout.

ios-review reviewed (changes requested):
Галерея картинок 🔥
Котики милые)

Все экраны с галерей можно объединить в один.
На этом экране будет просто несколько способов изменить текущую картинку.

И еще комментарии ниже)

TatyanaV24 committed:
Правки 26.01

Внесла правки на основании рекомендации на гитхаб.
Сейчас все задачи отображаются на одном контроллере.
Остался открытым вопросы по задаче с UIDatePicker, UITextField и UILabel.
Я правильно понимаю,  в UIDatePicker пользователь выбирает дату, в UITextField  вводит в формате цифры часовой пояс, в UILabel отображается дата уже с учетом часового пояса.  Немного не понимаю, какой должен получиться результат. Можете привести пример, чтобы я поняла, что именно необходимо сделать?
Также по UIStepper, настроики для изменения изображения при активации кнопки я добавила, но у меня отображаются только синие квадратики.

TatyanaV24 committed:
-

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/ViewController.swift:
```diff
@@ -19,22 +19,66 @@ class ViewController: UIViewController {
+        if index != Int(round(sender.value)) {
+            changeCatsImage()
+        }
+    }
```
При движении слайдера прогресс вью постоянно изменяет свое значение, даже если слайдер уже полностью передвинут направо.
Получается эффект вечного прогресса, который доходит до максимума, потом сбрасывается, а потом снова до максимума.

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/ViewController.swift:
```diff
@@ -19,22 +19,66 @@ class ViewController: UIViewController {
+        } else{
+            playImageView.alpha = 0.3
+            loadingActivity.startAnimating()
         }
```
<img width="425" alt="Screenshot 2021-02-13 at 21 32 55" src="https://user-images.githubusercontent.com/62285448/107852426-33737180-6e43-11eb-963a-f991332b36f7.png">

switch и stepper не нажимаются на маленьких экранах.

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/ViewController.swift:
```diff
@@ -19,22 +19,66 @@ class ViewController: UIViewController {
     @IBOutlet weak var playButton: UIButton!
+    @IBOutlet weak var blueSlider: UISlider!
+    @IBOutlet weak var customStepper: UIStepper!
+    @IBOutlet weak var progressView: UIProgressView!
```
Ошибка layout
<img width="1414" alt="Screenshot 2021-02-13 at 21 33 42" src="https://user-images.githubusercontent.com/62285448/107852438-45edab00-6e43-11eb-93df-34377fa0a03b.png">

ios-review reviewed (changes requested):
Классно, что решили объединить в один экран несколько заданий👍

Некоторое поведение требует правок.

Задание с календарем работает следующим способом:
1. Пользователь выбирает дату.
2. Эта дата отображается в лейбле.
3. Пользователь вводит часовой пояс в виде числа. Это число должно быть ограничено, так как часовые пояса ограничены.
4. В лейбле отображается дата + время, поправленные на введенный часовой пояс.

TatyanaV24 committed:
Календарь + Слайдер

Внесла правки по задаче со слайдером, сейчас все должно работать корректно.
По задаче с DataPicker и часовыми поясами, вроде решение есть, но оно мне не нравится, т.к. работает не на 100%. как будто не хватает кнопки на экране пользователя или второго лейбла....Еще была математическая дилемма, как именно осуществлять расчет при указании часового пояса. Я взяла за основу московское время и с учетом изменения пояса прибавляется или отнимаются часы. Получилось довольно детское решение. Мне кажется я что-то упустила в решении этого задания. Буду признательна за комментарии и идеи, на что мне стоит обратить внимание.
Спасибо.

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/CalendarViewController.swift:
```diff
@@ -0,0 +1,45 @@
+                dateLabel.text = "\(day)-\(month)-\(year) Time \(newHour):\(minute):\(second)"
+                
+            }
+        }
```
Попробуем такую реализацию:
```
    @IBAction func changeLabel(_ sender: UIDatePicker) {
        let pickerDate = sender.date
        let calendar = Calendar.current
        
        let timezoneValue = Int(newTimeTextField.text ?? "") ?? 0
        let endDate = calendar.date(byAdding: .hour, value: timezoneValue, to: pickerDate)
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "dd-MM-yyyy"// 'Time' HH:mm:ss"
        dateLabel.text = dateFormatter.string(from: endDate)
    }
```

В этой реализации нужно добавить вывод времени в dateFormat.
И будет здорово, если добавите возможность ввода отрицательных временных зон.
И еще валидацию введенных данных.
Не любое число является валидным часовым поясом.
<img width="1490" alt="Screenshot 2021-02-18 at 14 01 57" src="https://user-images.githubusercontent.com/62285448/108318172-f5cf6980-71f1-11eb-861e-e9f2a624994a.png">

ios-review replied:
При некорректных введенных данных требуется показать сообщение об ошибке.

TatyanaV24 replied:
Сразу хочу уточнить, правильно ли я понимаю эту строку коду - let timezoneValue = Int(newTimeTextField.text ?? "") ?? 0 -  timezoneValue присваиваем значение Int(newTimeTextField.text) если оно не пустое и не равно нулю.
Если в строке полностью раскрыть dateFormatter, т.е. dateFormatter.dateFormat = "dd-MM-yyyy 'Time' HH:mm:ss", то в лейбл отобразить выбранная пользователем дата и время компьютера, при введение в текст филд числа, чтобы произошел пересчет времени, необходимо вновь "крутить барабан". В моем представлении это неверная реализация, должно работать следующим образом, выбирается дата, которая отображается в лейбл. Далее в текстфилд вводим часовой пояс и после этого в лейбле должно отобразиться дата и время с учетом часового пояса. Как это осуществить, я не смогла придумать.
Честно говоря я не очень понимаю, что именно необходимо, добавить, чтобы "В этой реализации нужно добавить вывод времени в dateFormat." - если я правильно оцениваю, то должна быть еще одна строка, кода которая выводит новые данные в полном формате с поправкой на часовой пояс.
Пожалуйста, объясните подробнее, а том не могу понять(

ios-review replied:
Значение лейбла необходимо менять после двух событий:
1. Пользователь выбрал другое время в UIDatePicker (value changed).
2. Пользователь ввел новое значение в UITextField (editing changed).

После обоих событий необходимо выполнить одинаковый код, то есть функцию changeLabel.
Удалите параметр у этой функции:
`@IBAction func changeLabel() {}`
 А затем из сториборда назначьте два события на эту функцию.

ios-review reviewed (changes requested):
По слайдеру 👌

Реализация часовых поясов можно упростить.
Плюс добавил еще некоторые требования.

TatyanaV24 committed:
.

ios-review reviewed (changes requested):
Про изменение лейбла ответил в комментарии выше.
Нужно добавить еще валидацию введенных данных.

TatyanaV24 committed:
заключительные правки

Внесла необходимые изменения. Сейчас все работает верно.

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/CalendarViewController.swift:
```diff
@@ -13,19 +13,16 @@ class CalendarViewController: UIViewController {
         
-        let pickerDate = sender.date
+        let pickerDate = calendarDate.date
         let calendar = Calendar.current
```
🔥

ios-review commented on Additional-tasks-for-module-8/Additional-tasks-for-module-8/CalendarViewController.swift:
```diff
@@ -13,19 +13,16 @@ class CalendarViewController: UIViewController {
+        let pickerDate = calendarDate.date
         let calendar = Calendar.current
-        
         let timezoneValue = Int(newTimeTextField.text ?? "") ?? 0
```
Попробуете познакомиться с понятием валидация данных?

Валидация данных - проверка соответствия данных определенным условиям.

Внутри системы стоит валидировать (проверять) те данные, которые поступают в систему из вне. 
Валидаторы проверяют соответствие входного значения или же результата от другой системы определенному допустимому множеству значений для нашей системы. 

Пользователь в текстовое поле может ввести любое значение, даже простую строку.
Сейчас поведение в случае некорректного ввода пользователем упрощено до взятия нуля в качестве time zone.
Такое поведение 👌, так как ноль нейтрален по отношению к сложению.

Пользователи любят "отзывчивые, интерактивные" приложения, которые умеют сообщать более подробную информацию о его действиях.
Пользователь был бы доволен, если бы введенное значение в текстовое поле проверялось на соответствие диапазону чисел -12 ... 14.
В случае соответствия расчет и вывод новой даты.
Иначе вывод в лейбл (в тот же, куда и дату) текста ошибки о некорректности введенных данных.

dateLabel стоит переименовать на resultLabel, так как сейчас в него выводится уже два разных типа текста: дата и ошибка.
Или можно добавить еще один лейбл errorLabel.
В случае ошибки показывать errorLabel и скрывать dateLabel.
В случае новой даты наоборот.

ios-review reviewed (approved):
🚀

Задание с валидацией на ваше усмотрение)

TatyanaV24 committed:
Добавила условие валидации
